File: example/nodejs/eva.js
const fsPromise = require("fs").promises
const uuidv1 = require("uuid").v1
const EduSigner = require("./Signer")
const request = require("request")

class EvaClient{
    constructor(appId, appSecret,audioBuffer, sText){
        this.appId = appId
        this.appSecret = appSecret
        this.iFrameSize = 6400
        this.iMaxSeq = 999999
        this.sText = sText
        this.audioBuffer = audioBuffer
        this.schema = "https"
        this.method = "post"
        this.host = "smartedu.html5.qq.com"
        this.path = "/v1/sentence_follow"
    }

    async sendFrame(){
        let iOffset = 0, iSeq = 0, cPosBits = 2
        let uuid = uuidv1(), finalRsp = {}
        while(iOffset < this.audioBuffer.length && iSeq <= this.iMaxSeq){
            let frameBuffer = this.audioBuffer.slice(iOffset, iOffset+this.iFrameSize)
            if(iOffset == 0){
                cPosBits = 1
            } else if (iOffset + this.iFrameSize >= this.audioBuffer.length || iSeq == this.iMaxSeq){
                cPosBits = 4
            } else {
                cPosBits = 2
            }
        
            let start = new Date().getTime()
            let stRsp = await new Promise((resolve, reject)=>{
                let body = {
                    sText:this.sText,
                    sAudio: frameBuffer.toString("base64"),
                    iAudioFormat: 0x92,
                    sSessionId: uuid,
                    iSeq:iSeq,
                    cPosBits: cPosBits,
                    iOffset:iOffset,
                    sGuid:"test-guid"
                }
                let timestamp = new Date().getTime()
                let signer = new EduSigner(this.appId,this.appSecret,timestamp, this.method,this.host,this.path,{},body)
                request.post(`${this.schema}://${this.host}${this.path}`, {
                    form: body,
                    headers:{
                        Authorization:`algorithm=sha256&appid=${this.appId}&timestamp=${timestamp}&sig=${signer.signature}`
                    },
                    json: true
                }, (err, response)=>{
                    if(err){
                        reject(err)
                    } else {
                        console.log(response.body)
                        resolve(response)
                    }
                })
            })
            if(stRsp.body.iEvaStatus == 0){
                finalRsp = stRsp.body
            }
            iSeq++
            iOffset += this.iFrameSize
            await this.sleep(200 - (new Date().getTime() - start))
        }
        return finalRsp
    }
    sleep(t){
        t = t>0?t:0
        return new Promise((resolve, reject)=>{
            setTimeout(resolve, t)
        })
    }
}

async function test(){
    try{
        const audioBuffer = await fsPromise.readFile("test.wav")
        asrClient = new EvaClient("your appid", "your appkey", audioBuffer,"how are you")
        let response = await asrClient.sendFrame()
        console.log(response)
    } catch(e){
        console.log("send error:", e)
    }
    
}

test()

File: example/python/eva.py
import sys, time, uuid, base64
import requests
from Signer import Signer 

class EvaClient:
    def __init__(self, appId, appSecret, audioBuffer, sText, session):
        self.appId = appId
        self.appSecret = appSecret
        self.iFrameSize = 6400
        self.audioBuffer = audioBuffer
        self.schema = "https"
        self.method = "post"
        self.host = "smartedu.html5.qq.com"
        self.path = "/v1/word_follow"
        self.sText = sText
        self.session = session

    def send(self):
        iOffset, iSeq, cPosBits, uuidv1 = 0, 0, 2, uuid.uuid1()
        allLen = len(self.audioBuffer)
        while iOffset < allLen:
            timestamp = int(round(time.time() * 1000))
            frameBuffer = self.audioBuffer[iOffset:iOffset + self.iFrameSize]
            frameBase64 = base64.b64encode(frameBuffer).decode()
            if iOffset == 0:
                cPosBits = 1
            elif (iOffset+self.iFrameSize) < allLen:
                cPosBits = 2
            else:
                cPosBits = 4
            bodyParams = {
                "sAudio": frameBase64,
                "sSessionId": uuidv1,
                "iSeq": iSeq,
                "cPosBits": cPosBits,
                "iOffset":iOffset,
                "iAudioFormat": 146,
                "sText":self.sText,
                "sGuid":"test-guid"
            }
            signer = Signer(self.appId, self.appSecret, timestamp, self.method, self.host, self.path, {}, bodyParams )
            authHeader = signer.getAuthorizationHeader()
            response = self.session.post(url="{}://{}{}".format(self.schema, self.host, self.path),data=bodyParams,headers={"Authorization": authHeader})
            now = int(round(time.time() * 1000))
            diff = now - timestamp
            print("{} {}/{}".format(diff, iOffset, allLen), len(frameBase64), bodyParams["iSeq"],cPosBits, response.elapsed.microseconds, response.text)
            iSeq += 1
            iOffset += self.iFrameSize
            gap = 60
            if(diff < gap):
                time.sleep((gap-diff)/1000)

if __name__ == "__main__":
    # 注意：使用此方法以开启http的keep-alive选项，能避免多次建立、断开tcp连接的开销，大幅提升接口性能
    # http的keep-alive，各语言实现方式不同，但都应该开启
    session = requests.session()
    with open("./test.wav", "rb") as audioBuffer:
         asrClient = EvaClient("you appid", "your appsecret", audioBuffer.read(),"how are you", session)
         asrClient.send()